pkg/controller/persistentvolume/persistentvolume_provisioner_controller.go

```diff
@@ -0,0 +1,489 @@
+	"k8s.io/kubernetes/pkg/types"
+	"k8s.io/kubernetes/pkg/util/io"
+	"k8s.io/kubernetes/pkg/util/mount"
+	"k8s.io/kubernetes/pkg/volume"
+	"k8s.io/kubernetes/pkg/watch"
+
+	"github.com/golang/glog"
+)
+
+// PersistentVolumeProvisionerController reconciles the state of all PersistentVolumes and PersistentVolumeClaims.
+type PersistentVolumeProvisionerController struct {
+	volumeController *framework.Controller
+	volumeStore      cache.Store
+	claimController  *framework.Controller
+	claimStore       cache.Store
+	client           controllerClient
+	cloud            cloudprovider.Interface
+	provisioner      volume.ProvisionableVolumePlugin
+	pluginMgr        volume.VolumePluginMgr
+	stopChannels     map[string]chan struct{}
+	mutex            sync.RWMutex
+}
+
+// NewPersistentVolumeProvisionerController creates a new PersistentVolumeProvisionerController
+func NewPersistentVolumeProvisionerController(client controllerClient, syncPeriod time.Duration, plugins []volume.VolumePlugin, provisioner volume.ProvisionableVolumePlugin, cloud cloudprovider.Interface) (*PersistentVolumeProvisionerController, error) {
+	controller := &PersistentVolumeProvisionerController{
+		client:      client,
+		cloud:       cloud,
+		provisioner: provisioner,
+	}
+
+	if err := controller.pluginMgr.InitPlugins(plugins, controller); err != nil {
+		return nil, fmt.Errorf("Could not initialize volume plugins for PersistentVolumeProvisionerController: %+v", err)
+	}
+
+	glog.V(5).Infof("Initializing provisioner: %s", controller.provisioner.Name())
+	controller.provisioner.Init(controller)
+
+	controller.volumeStore, controller.volumeController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options unversioned.ListOptions) (runtime.Object, error) {
+				return client.ListPersistentVolumes(options)
+			},
+			WatchFunc: func(options unversioned.ListOptions) (watch.Interface, error) {
+				return client.WatchPersistentVolumes(options)
+			},
+		},
+		&api.PersistentVolume{},
+		syncPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    controller.handleAddVolume,
+			UpdateFunc: controller.handleUpdateVolume,
+			// delete handler not needed.
+			// 			DeleteFunc: controller.handleDeleteVolume,
+		},
+	)
+	controller.claimStore, controller.claimController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options unversioned.ListOptions) (runtime.Object, error) {
+				return client.ListPersistentVolumeClaims(api.NamespaceAll, options)
+			},
+			WatchFunc: func(options unversioned.ListOptions) (watch.Interface, error) {
+				return client.WatchPersistentVolumeClaims(api.NamespaceAll, options)
+			},
+		},
+		&api.PersistentVolumeClaim{},
+		syncPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    controller.handleAddClaim,
+			UpdateFunc: controller.handleUpdateClaim,
+			// delete handler not needed.
+			// normal recycling applies when a claim is deleted.
+			// it requires no provisioning or processing by this controller.
```

Comment:
Might be more helpful to say that these delete events are processed by the claim binder controller.